tiledb/sm/query/sparse_global_order_reader.cc

```diff
@@ -0,0 +1,918 @@
+
+  // Remove processed data from memory.
+  RETURN_NOT_OK(clean_up_processed_data(to_copy));
+
+  return Status::Ok();
+}
+
+void SparseGlobalOrderReader::reset() {
+}
+
+Status SparseGlobalOrderReader::load_initial_data() {
+  if (initial_data_loaded_)
+    return Status::Ok();
+
+  // For easy reference.
+  auto fragment_num = fragment_metadata_.size();
+
+  // Calculate a bit vector of the tiles in the subarray, if set.
+  if (subarray_.is_set()) {
+    // TODO Compute the memory usage here and release after computation.
+    RETURN_NOT_OK(subarray_.precompute_tile_overlap(
+        0, 0, &config_, storage_manager_->compute_tp()));
+
+    // TODO Warning against memory budget if this gets too big and add a
+    // configuration option to turn off.
+    result_tile_set_.resize(fragment_num);
+    RETURN_CANCEL_OR_ERROR(
+        compute_result_tiles_set(subarray_, result_tile_set_));
+  }
+
+  // Preload zipped coordinate tile offsets. Note that this will
+  // ignore fragments with a version >= 5.
+  RETURN_CANCEL_OR_ERROR(load_tile_offsets(subarray_, {constants::coords}));
+
+  // Preload unzipped coordinate tile offsets. Note that this will
+  // ignore fragments with a version < 5.
+  const auto dim_num = array_schema_->dim_num();
+  dim_names_.reserve(dim_num);
+  is_dim_var_size_.reserve(dim_num);
+  for (unsigned d = 0; d < dim_num; ++d) {
+    dim_names_.emplace_back(array_schema_->dimension(d)->name());
+    is_dim_var_size_[d] = array_schema_->var_size(dim_names_[d]);
+  }
+  RETURN_CANCEL_OR_ERROR(load_tile_offsets(subarray_, dim_names_));
+
+  // Make sure we have enough space for tiles data.
+  result_tiles_.resize(fragment_num);
+  frag_tile_idx_.resize(fragment_num);
+  all_tiles_loaded_.resize(fragment_num);
+  memory_used_for_coords_.resize(fragment_num);
+
+  initial_data_loaded_ = true;
+  return Status::Ok();
+}
+
+Status SparseGlobalOrderReader::get_coord_tile_size(
+    unsigned dim_num, unsigned f, uint64_t t, uint64_t* tile_size) {
+  for (unsigned d = 0; d < dim_num; d++) {
+    *tile_size += fragment_metadata_[f]->tile_size(dim_names_[d], t);
+
+    if (is_dim_var_size_[d]) {
+      uint64_t temp = 0;
+      RETURN_NOT_OK(fragment_metadata_[f]->tile_var_size(
+          *encryption_key_, dim_names_[d], t, &temp));
+      *tile_size += temp;
+    }
+  }
+
+  return Status::Ok();
+}
+
+Status SparseGlobalOrderReader::load_coord_tile(
+    unsigned dim_num,
+    uint64_t per_fragment_memory,
+    unsigned f,
+    uint64_t t,
+    const Domain* domain,
+    bool* budget_exceeded) {
+  // Calculate memory consumption for this tile.
+  uint64_t tile_size = 0;
+  RETURN_NOT_OK(get_coord_tile_size(dim_num, f, t, &tile_size));
+
+  // Don't load more tiles than the memory budget.
+  if (memory_used_for_coords_[f] + tile_size > per_fragment_memory) {
+    *budget_exceeded = true;
+    return Status::Ok();
+  }
+
+  memory_used_for_coords_[f] += tile_size;
+
+  result_tiles_[f].emplace_back(f, t, domain);
```

Comment:
Now I see that this function does not load tiles - we need to be clearer in the name. For example, `add_result_tile` or something like that.